### PR TITLE
Fix #1809 by avoiding making unnecessary copies when Tables signals t…

### LINKED ISSUE #1809: Unnecessary copy for any Queryverse source
I think (if I understand the current state of constructors and Tables.jl correctly) that we have a pretty major regression for the entire Queryverse integration with the recent versions because essentially anything a la:
```julia
df = DataFrame(anything_queryverse)
```
makes an unnecessary temporary copy of all the columns during the construction of the `DataFrame` instance.

I think here is what is happening currently: 

a) if `anything_queryverse` implements the `get_columns_copy_using_missing` interface from [TableTraits.jl](https://github.com/queryverse/TableTraits.jl/blob/816661ce6c5e16e92a161a7fa35201c597a023ad/src/TableTraits.jl#L35), then `Tables.columns` will return the column vectors that `TableTraits.get_columns_copy_using_missing ` returned, and then the `DataFrame` constructor will make a copy of each of these columns and use those. That is not good, because the whole point of the `get_columns_copy_using_missing` interface is that the source allocates a set of column vectors that the sink then owns, i.e. they are by the definition of the interface already copies that the sink can use.

b) if `anything_queryverse` implements the iterable tables interface (i.e an iterator of named tuples), then `Tables.columns` will allocate new column vectors and populate them with the values from the source, and then return those. And then the `DataFrame` constructor makes copies of those vectors, which again are really unnecessary large temporary allocations. After all, those vectors that are being allocated are _only_ allocated for this specific sink, so really they are best interpreted as copies for the sink as well.

I'm not sure what a good solution to this is, but it would be good to find some solution to this, given that this used to work well for a long time before the recent set of changes.

It would of course be easy to solve this if DataFrames.jl implemented the TableTraits.jl interfaces directly: they are explicit about column ownership, so the `DataFrame` constructor would always know whether it is getting columns from a source that it (DataFrame) then owns. I think the difficulty here stems from the fact that `Tables.columns` does not distinguish between a case where a source is returning a copy of columns and one where it is returning something that is still owned by the source. So what happens in the end is that `Table.columns` when called on any TableTraits.jl source will return vectors that are already copies, but for other sources `Tables.columns` can return a view. As the consumer, the `DataFrame` constructor then doesn't really know what to do: in the first case it shouldn't make a copy of what is returned by `Tables.columns`, in the second it should. I think the core problem is that those two cases are now conflated in the `Tables.columns` function.

@quinnj any idea what one could do about this?

EDIT: Oh, and I should add that I might well misunderstand what is happening right now! So a sanity check whether my description of when copies are made is correct would also be great!

### PR DESCRIPTION
**Pull request by quinnj**
…hat the incoming source has already been copied

cc: @davidanthoff @bkamins @nalimilan 

Relies on https://github.com/JuliaData/Tables.jl/pull/94

### CODE REVIEW

**Comment by bkamins** [comment]
Looks good. We should merge it after https://github.com/JuliaData/Tables.jl/pull/94 is released, unless the design there is changed following what @davidanthoff commented there.

**Comment by davidanthoff** [comment]
I think I would prefer that we fix this first (so merge https://github.com/JuliaData/Tables.jl/pull/94 and this PR here and tag). The other comment I made over there is really a larger, meta issue, and I think it would be better if that didn't hold up a fix.

File: src/other/tables.jl [hunk 1]
```diff
@@ -11,6 +11,10 @@ Tables.materializer(df::AbstractDataFrame) = DataFrame
 
 getvector(x::AbstractVector) = x
 getvector(x) = collect(x)
+fromcolumns(x::Tables.CopiedColumns; kw...) =
```
Comment (nalimilan) [inline src/other/tables.jl]: Better not accept any keyword argument but ignore them: 
```suggestion
fromcolumns(x::Tables.CopiedColumns; copycols::Bool=true) =
```

Comment (quinnj) [reply src/other/tables.jl]: Why? It's an internal function and I was just trying to emphasize that `copycols` is ignored in this case.

Comment (nalimilan) [reply src/other/tables.jl]: Well it's a matter of taste, but `kw...` isn't very explicit.

**Comment by quinnj** [comment]
Ok, I think this is good to go; not sure why appveyor check got messed up here, but the results for the most recent commit are all green: https://ci.appveyor.com/project/nalimilan/dataframes-jl/builds/24597257

**Comment by bkamins** [comment]
@davidanthoff if we merge this do we need https://github.com/JuliaData/DataFrames.jl/pull/1810?

**Comment by davidanthoff** [comment]
>  if we merge this do we need #1810?

No, I think with this we probably don't need #1810.

File: src/other/tables.jl (hunk 1 above)
Comment (quinnj) [reply src/other/tables.jl]: kk, updated the keyword arg and added a comment that `copycols` is ignored for `CopiedColumns`

**Comment by davidanthoff** [comment]
Thanks for sorting this out!

**Pull request merged**